blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java

```diff
@@ -206,4 +207,60 @@ public static boolean isAgentStart(@Nullable FlowNode node) {
 
         return false;
     }
+
+    /**
+     * Gives top level stage or parallel node. A top level stage or parallel is one that doesn't have another stage or
+     *  parallel between it and FlowStartNode
+     * @param node it could be any node inside a stage or parallel
+     * @return Immediate stage or parallel node
+     */
+    public static @CheckForNull FlowNode getEnclosingStageOrParallel(@Nonnull FlowNode node) {
+        FlowNode stageOrParallel = getStageOrParallelParent(node);
+        if(stageOrParallel != null) {
+            return getEnclosingStageOrParellel(stageOrParallel, stageOrParallel.getParents());
+        }
+        return null;
+    }
+
+    /**
+     * Gives immediate enclosing Stage or parallel node.
+     * @param node it could be any node inside a stage or parallel
+     * @return Immediate stage or parallel node
+     */
+    private static @CheckForNull FlowNode getStageOrParallelParent(@Nonnull FlowNode node) {
```

Comment:
🐛 No, this will generate a very uncontrolled StackOverflowException in real-world use cases and extremely bad big-O scaling -- please use: https://github.com/jenkinsci/workflow-api-plugin/pull/50 `iterateEnclosingBlocks(FlowNode)` -- that's what it's there for. :)   Also, use a Predicate to match Stage/Parallel -- probably https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/NodeStepTypePredicate.java